x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -141,4 +202,127 @@ public void removeFromCache(final Predicate<CacheKey> predicate) {
+    void onCacheFileUpdate(CacheFile cacheFile) {
+        assert cacheFile != null;
+        cacheFilesToSync.offer(cacheFile);
+        numberOfCacheFilesToSync.incrementAndGet();
+    }
+
+    /**
+     * This method is invoked after a {@link CacheFile} is evicted from the cache.
+     * <p>
+     * It notifies the {@link CacheFile}'s eviction listeners that the instance is evicted.
+     *
+     * @param cacheFile the evicted instance
+     */
+    void onCacheFileRemoval(CacheFile cacheFile) {
+        IOUtils.closeWhileHandlingException(cacheFile::startEviction);
+    }
+
+    // used in tests
+    boolean isCacheFileToSync(CacheFile cacheFile) {
+        return cacheFilesToSync.contains(cacheFile);
+    }
+
+    /**
+     * Synchronize the cache files and their parent directories on disk.
+     *
+     * This method synchronizes the cache files that have been updated since the last time the method was invoked. To be able to do this,
+     * the cache files must notify the {@link CacheService} when they need to be fsync. When a {@link CacheFile} notifies the service the
+     * {@link CacheFile} instance is added to the current queue of cache files to synchronize referenced by {@link #cacheFilesToSync}.
+     *
+     * Cache files are serially synchronized using the {@link CacheFile#fsync()} method. When the {@link CacheFile#fsync()} call returns a
+     * non empty set of completed ranges this method also fsync the shard's snapshot cache directory, which is the parent directory of the
+     * cache entry. Note that cache files might be evicted during the synchronization.
+     */
+    protected synchronized void synchronizeCache() {
+        long count = 0L;
+        final Set<Path> cacheDirs = new HashSet<>();
+        final long startTimeNanos = threadPool.relativeTimeInNanos();
+        final long maxCacheFilesToSync = Math.min(numberOfCacheFilesToSync.get(), this.maxCacheFilesToSyncAtOnce);
+        for (long i = 0L; i < maxCacheFilesToSync; i++) {
+            if (lifecycleState() != Lifecycle.State.STARTED) {
+                logger.debug("stopping cache synchronization (cache service is closing)");
+                break;
+            }
+            final CacheFile cacheFile = cacheFilesToSync.poll();
+            assert cacheFile != null;
+
+            final long value = numberOfCacheFilesToSync.decrementAndGet();
+            assert value >= 0 : value;
+            final Path cacheFilePath = cacheFile.getFile();
+            try {
+                final SortedSet<Tuple<Long, Long>> ranges = cacheFile.fsync();
+                if (ranges.isEmpty() == false) {
+                    logger.trace("cache file [{}] synchronized with [{}] completed range(s)", cacheFilePath.getFileName(), ranges.size());
+                    final Path cacheDir = cacheFilePath.toAbsolutePath().getParent();
+                    if (cacheDirs.add(cacheDir)) {
+                        try {
+                            IOUtils.fsync(cacheDir, true, false);
```

Comment:
I found it simpler to do it here and concentrate the fsync logic at one place, but I agree it means an extra fsync per cache directory every 60s. 

> Perhaps we should do that when creating the file instead?

That would mean to fsync the cache dir the first time a cache file is opened and have a mechanism to keep the knowledge of which cache dir was already fsync or not, since the cache file can be evicted before the cache sync task is executed. That sounds a bit more complex for a little gain to me.